control-plane-operator/hostedclusterconfigoperator/controllers/drainer/drainer.go

```diff
@@ -0,0 +1,162 @@
+package drainer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/openshift/hypershift/support/releaseinfo"
+	"github.com/openshift/hypershift/support/upsert"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/kubectl/pkg/drain"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	// DesiredDrainerAnnotationKey is set by the MCD to indicate drain/uncordon requests
+	DesiredDrainerAnnotationKey = "machineconfiguration.openshift.io/desiredDrain"
+	// LastAppliedDrainerAnnotationKey is set by the controller to indicate the last request applied
+	LastAppliedDrainerAnnotationKey = "machineconfiguration.openshift.io/lastAppliedDrain"
+	// DrainerStateUncordon is used for drainer annotation as a value to indicate needing an uncordon
+	DrainerStateUncordon = "uncordon"
+	// DrainerStateDrain = "drain"
+	DrainerStateDrain = "drain"
+	// drainFailureRequeueDuration is the delay until we retry draining the node, such that other nodes
+	// get a chance in the queue
+	drainFailureRequeueDuration = 1 * time.Minute
+)
+
+type Reconciler struct {
+	client                 client.Client
+	guestClusterClient     client.Client
+	guestClusterKubeClient kubeclient.Interface
+	releaseProvider        releaseinfo.Provider
+	upsert.CreateOrUpdateProvider
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Reconciling")
+
+	node := &corev1.Node{}
+	err := r.guestClusterClient.Get(ctx, req.NamespacedName, node)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("not found", "request", req.String())
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "error getting node")
+		return ctrl.Result{}, err
+	}
+
+	if node.Annotations[DesiredDrainerAnnotationKey] != node.Annotations[LastAppliedDrainerAnnotationKey] {
```

Comment:
nit: I'd rather return early if node.Annotations[DesiredDrainerAnnotationKey] == node.Annotations[LastAppliedDrainerAnnotationKey] or even have a predicate to filter out the event than having the whole main reconciliation logic within a if block. Feel free to keep it as it's, just wanted to point out the alternatives.